### PR TITLE
add pgfplot

### PR DESCRIPTION
**Pull request by ksagiyam**
# Description

Enable outputting `Function` data for LaTeX tikz plots. Not for large `Function`s.

Example:

```
from firedrake import *
  

mesh = TorusMesh(31, 13, 2.0, 0.5, quadrilateral=False, print_latex_example=True)
x, y, z = SpatialCoordinate(mesh)
V = FunctionSpace(mesh, "CG", 2)
f = Function(V, name="torus_tri").interpolate(x)
pgfplot(f, "./pgfplot_example_torus_tri.txt", degree=2)
```
generates an output file "./pgfplot_example_torus_tri.txt" and prints an example usage in LaTeX:
```
===========================================================================================
% pgfplot_example.tex

\documentclass{article}
\usepackage{tikz}
\usepackage{pgfplots}
\usepgfplotslibrary{patchplots}
\pgfplotsset{compat=1.18}
\begin{document}
\begin{figure}[ht]
\begin{tikzpicture}
\begin{axis}[title=torus tri,
             xmin=-2.49, xmax= 2.50,
             ymin=-2.50, ymax= 2.50,
             zmin=-0.50, zmax= 0.50,
             xlabel={$x$},
             ylabel={$y$},
             zlabel={$z$},
             % xtick={0, 1, 2},
             % xticklabels={0, 1, 2},
             % ytick={0, 1},
             % yticklabels={0, 1},
             % ztick={0, 1},
             % zticklabels={0, 1},
             % axis equal,
             axis equal image,
             colorbar,
             colormap/hot, % hot, cool, bluered, greenyellow, redyellow, violet, blackwhite
             colorbar/width=10pt,
             view={30}{45},
             width=300pt,
             height=300pt,
             % axis line style={draw=none},
             % tick style={draw=none},
            ]
\addplot3[patch,
          patch type=triangle quadr,
          point meta=explicit,
          shader=faceted interp, % interp, faceted, faceted interp
          opacity=1.,
         ] table[x=x, y=y, z=z, meta=torus_tri] {./pgfplot_example_torus_tri.txt};
\end{axis}
\end{tikzpicture}
\end{figure}
\end{document}
===========================================================================================

Run:

lualatex -shell-escape pgfplot_example.tex

For more details, see:

https://pgfplots.sourceforge.net/gallery.html
http://www.bakoma-tex.com/doc/latex/pgfplots/pgfplots.pdf

```

## Fixes Issues:

# Checklist for author:

- [ ] I have linted the codebase (`make lint` in the `firedrake` source directory).
- [ ] My changes generate no new warnings.
- [ ] All of my functions and classes have appropriate docstrings.
- [ ] I have commented my code where its purpose may be unclear.
- [ ] I have included and updated any relevant documentation.
- [ ] Documentation builds locally (`make linkcheck; make html; make latexpdf` in `firedrake/docs` directory)
- [ ] I have added tests specific to the issues fixed in this PR.
- [ ] I have added tests that exercise the new functionality I have introduced
- [ ] Tests pass locally (`pytest tests` in the `firedrake` source directory) (useful, but not essential if you don't have suitable hardware).
- [ ] I have performed a self-review of my own code using the below guidelines.

# Checklist for reviewer:

- [ ] Docstrings present
- [ ] New tests present
- [ ] Code correctly commented
- [ ] No bad "code smells"
- [ ] No issues in parallel
- [ ] No CI issues (excessive parallelism/memory usage/time/warnings generated)
- [ ] Upstream/dependent branches and PRs are ready

Feel free to add reviewers if you know there is someone who is already aware of this work.

Please open this PR initially as a draft and mark as ready for review once CI tests are passing.

### CODE REVIEW

**Comment by ksagiyam** [comment]
```
mesh = UnitBallMesh(refinement_level=3)
x, y, z = SpatialCoordinate(mesh)
V = FunctionSpace(mesh, "CG", 2)
f = Function(V, name="ball").interpolate(x)
pgfplot(f, "./pgfplot_example_ball.txt", degree=2)
```
[ball.pdf](https://github.com/firedrakeproject/firedrake/files/11519188/ball.pdf)

**Comment by ksagiyam** [comment]
```
mesh = BoxMesh(32, 16, 16, 2.0, 1.0, 1.0, hexahedral=True)
x, y, z = SpatialCoordinate(mesh)
V = FunctionSpace(mesh, "CG", 2)
r = sqrt(x**2 + y**2 + z**2)
f = Function(V, name="box").interpolate(sin(16 * r) / (r + 1.e-6))
pgfplot(f, "./pgfplot_example_box.txt", degree=2)
```
[box.pdf](https://github.com/firedrakeproject/firedrake/files/11519189/box.pdf)

File: firedrake/pgfplot.py
Comment (connorjward) [inline firedrake/pgfplot.py:1]: Surely this should be part of `firedrake.plot`? Perhaps we should make `firedrake.plot` a subpackage. If we are careful with namespacing this shouldn't affect an end user at all.

**Review by connorjward — COMMENTED** [review]
Cool!

File: firedrake/pgfplot.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+        perms = map_facet_dofs[_facets.local_facet_dat.data_ro[:nfacets]]
+    else:
+        raise NotImplementedError(f"Got unsupported FIAT cell: {fiat_cell}")
+    patches_c = _pgfplot_create_patch_arrays(cdata, coordV.cell_node_list, cells, perms)
+    patches_f = _pgfplot_create_patch_arrays(fdata.reshape(-1, 1), V.cell_node_list, cells, perms)
+    patches = np.concatenate([patches_c, patches_f], axis=1)
+    return patches, patch_type
+
+
+def pgfplot(f, filename, degree=1, complex_component='real', print_latex_example=True):
```
[the remark this answers is not in the document]
Comment (ksagiyam) [reply firedrake/pgfplot.py:158]: Yes. Updated docstring.

File: firedrake/pgfplot.py
Comment (ksagiyam) [reply firedrake/pgfplot.py:1]: Thanks! Yes, we can.

File: firedrake/plot.py [hunk 2]
```diff
@@ -873,3 +877,295 @@ def __call__(self, function):
...
+    f
+       `Function` to plot.
+    filename
+        Name of the output file.
+    degree
+        Degree of interpolation for plotting: ``1`` (linear) or ``2`` (quadratic).
+    complex_component
+        Complex component to be plotted: 'real' or 'imag'.
+    print_latex_example
+        Flag indicating whether to print a latex example or not.
```
Comment (dham) [inline firedrake/plot.py:1045]: Types on parameters please.

**Review by dham — CHANGES_REQUESTED** [review]
Wow. Just wow.

File: firedrake/plot.py (hunk 2 above)
Comment (ksagiyam) [reply firedrake/plot.py:1045]: Fixed! Also added one paragraph in the `Notes`.

**Pull request merged**